[FSDP] Enable mixed hybrid/non-hybrid sharding strategies

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90862 [FSDP][3/N] Move `fsdp_modules(root_only=True)` -> `_get_fsdp_root_states()`
* #90861 [FSDP][2/N] Move `fsdp_modules(root_only=False)` -> `_get_fsdp_states()`
* #90864 [FSDP][Easy] Rename `entry` -> `fsdp_module` to be more descriptive
* #90860 [FSDP][1/N] Add `_get_fsdp_states()`
* **#90846 [FSDP] Enable mixed hybrid/non-hybrid sharding strategies**
* #90859 [FSDP][Easy] Use `run_subtests` for hybrid shard test
* #90858 [FSDP][Easy] ufmt files
* #90840 [FSDP][BE] Remove `_module_to_handles`, `HandleConfig`; use term "fqn"; clarify docs

In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

Commit:
[FSDP] Enable mixed hybrid/non-hybrid sharding strategies

[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable mixed hybrid/non-hybrid sharding strategies"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable mixed hybrid/non-hybrid sharding strategies"


In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable mixed hybrid/non-hybrid sharding strategies"


In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -1,19 +1,17 @@
-from functools import partial
-from collections import Counter
 import sys
+from collections import Counter
```
Could we have formatting changes separated out and in a different PR? Formatting changes with logical changes make it harder for reviewers to understand the critical parts of the PR to review.

awgu replied:
Sorry about that. I separated it out. Eventually, we should try to have all developers use ufmt so that we can ufmt files right before pushing. Converging to the same formatter saves the uncertainty of how code should be formatted, and we chose ufmt since PyTorch recommends that.

rohan-varma replied:
Curious about how we will encourage this practice. `lintrunner` is today's automated tool that can run pre-commit and is enforced by PyTorch CI. 

Should we work with dev infra / CI folks to add ufmt to `lintrunner`? Without automation, enforcement of a linting standard is prone to breaking.

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -54,6 +52,7 @@ def patch_allreduce(new_allreduce):
     finally:
         dist.all_reduce = orig_ar
 
+
```
Admittedly nitpicky, but these sort of empty line changes also make the logical changes to the PR to review in my experience.

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -97,67 +97,58 @@ def test_raises_manual_wrap_hybrid_shard_when_none_policy(self):
-            model, process_group=(intra_pg, dist.new_group()), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+            model,
+            process_group=(intra_pg, dist.new_group()),
+            sharding_strategy=ShardingStrategy.HYBRID_SHARD,
```
nit: these formatting changes are another example where reviewers may need to read carefully to check if there are any logical changes to the test.

awgu replied:
Makes sense.

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
-                self.assertEqual(num_flat_params, cntr[orig_ar])
-                self.assertEqual(num_flat_params, cntr[orig_rs])
-                dist.barrier()
+        self.run_subtests(
```
Ideally refactors such as switching to use `run_subtests` would be moved into a separate PR as they're unrelated to the description given in the PR.

awgu replied:
Makes sense. I separated it into a different PR.

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy._HYBRID_SHARD_ZERO2,
+                ],
+                "init_mode": ["uniform", "mixed"],
```
Would it be better to have an enum / constant here instead of hardcoded strings?

awgu replied:
Yes, good point. I refactored to an enum.

rohan-varma commented on test/distributed/fsdp/test_fsdp_hybrid_shard.py:
```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
+                CUDAInitMode.CUDA_BEFORE,
+                fsdp_kwargs,
+            )
+        elif init_mode == "mixed":
```
It looks like the difference between "uniform" and "mixed" is whether we mix and match HSDP sharding strategies. Could we better name these to indicate so?

awgu replied:
I provided better names: `ALL_HYBRID_SHARD` and `MIXED_HYBRID_FULL_SHARD`.

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -37,37 +37,41 @@
+    inter_node_states = set()
+    for fsdp_module in root_module.fsdp_modules(root_module):
+        # TODO: Change this to handle's sharding strategy if we deprecate
+        # `ShardingStrategy` internally.
```
Do we have an issue tracking the discussions / decisions needed in order to deprecate `ShardingStrategy`?

awgu replied:
https://github.com/pytorch/pytorch/issues/90857

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -108,16 +112,17 @@ def _lazy_init(
             # as each one already uses the same process group, but can slightly save memory
             # since other FSDP units allreduce state can be garbage collected.
-            fsdp_module._inter_node_state = state._inter_node_state
+            assert inter_node_state is not None, "Implementation error"
```
nit: can we have something more descriptive than implementation error, possibly hinting towards what the underlying issue may be?

awgu replied:
Added a detailed explanation.

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -37,37 +37,41 @@
-                f"For {fsdp_root.sharding_strategy}, inter-node process groups do not match"
-            )
+    Returns:
+        Tuple[HookState]: One of the instances' inter-node state (does not
```
curious, I don't see `HookState` anywhere?

awgu replied:
Good catch. I recalled it being called that, but it is actually called `DefaultState`. We may want to revisit that name in the future.

rohan-varma reviewed (commented):
Apologies for the preemptive stamp - coming back for a full review :)

Commit:
Update on "[FSDP] Enable mixed hybrid/non-hybrid sharding strategies"


In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Enable mixed hybrid/non-hybrid sharding strategies"


In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

[ghstack-poisoned]

awgu commented:
I will fix test failures tomorrow.

Commit:
Update on "[FSDP] Enable mixed hybrid/non-hybrid sharding strategies"


In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
